core/src/main/java/org/infinispan/interceptors/impl/CacheWriterInterceptor.java

```diff
@@ -262,7 +262,11 @@ protected void processIterableBatch(InvocationContext ctx, PutMapCommand cmd, Pe
             .filter(filter)
             .map(key -> createMarshalledEntry(ctx, key))
             .iterator();
-      persistenceManager.writeBatchToAllNonTxStores(iterable, mode, cmd.getFlagsBitSet());
+
+      // Ignore empty batches
+      if (iterable.iterator().hasNext()) {
```

Comment:
Sorry, I can't see why this was causing issues yet. I thought we ensured that all stores can handle empty `Iterator`s ?

Reply:
We did, however introducing the `StoreUnavailableException` has caused another problem with empty iterables. In the [CacheWriterInterceptor](https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/interceptors/impl/CacheWriterInterceptor.java#L253-L254) we call `processIterableBatch` twice, which results in two calls `PersistenceManager::writeBatchToAllNonTxStores`. It is possible for the store availability to change between the first and second method call, which results in the second call throwing a `StoreUnavailableException`. If the Iterable is non-empty, this is fine as it's not possible to write the modification to the store, however if the Iterable is empty this exception should not be thrown.

Reply:
Gotcha, so it isn't that it is called twice, but rather because [PersistenceManagerImpl](https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java#L594) checks the availability without caring about if there are any entries to write. In that case I wonder if this is correct spot to fix this.

Reply:
Yes very true. I will update all of the the PersistenceManager methods so that if the iterable is empty, then we return before even attempting to acquire the readLock

Reply:
I guess this means we can also remove the guideline that stores have to support empty Iterators now as well.

Reply:
I've added a revert commit to the PR